src/main/java/org/kaazing/netx/ws/internal/io/WsOutputStream.java

```diff
@@ -40,6 +46,118 @@ public void write(int b) throws IOException {
     public void write(byte[] b, int off, int len) throws IOException {
         out.write(0x82);
 
+        encodePayloadLength(len);
+
+        byte[] mask = new byte[4];
+        random.nextBytes(mask);
+        out.write(mask);
+
+        byte[] masked = new byte[len];
+        for (int i = 0; i < len; i++) {
+            int ioff = off + i;
+            masked[i] = (byte) (b[ioff] ^ mask[i % mask.length]);
+        }
+
+        out.write(masked);
+    }
+
+    public void writeClose(int code, byte[] reason) throws IOException {
+        if (closeSent) {
+            // If the client has already sent CLOSE, then it should NOT be sent again during the closing handshake when the
+            // server responds with a CLOSE.
+            return;
+        }
+
+        int len = 0;
+
+        if (code != 0) {
+            switch (code) {
+            case 1000:
+            case 1001:
+            case 1002:
+            case 1003:
+            case 1007:
+            case 1008:
+            case 1009:
+            case 1010:
+            case 1011:
+            case 1005:
+                len += 2;
+                break;
+            default:
+                if ((code >= 3000) && (code <= 4999)) {
+                    len += 2;
+                }
+
+                throw new IOException(format("Invalid CLOSE code %d", code));
+            }
+
+            if (reason != null) {
+//                if (reason.length > 123) {
+//                    throw new IOException("Reason must not be more than 123 bytes");
+//                }
+
+                len += reason.length;
+            }
+        }
+
+        out.write(0x88);
+
+        encodePayloadLength(len);
+
+        if (len == 0) {
+            out.write(EMPTY_MASK);
+        }
+        else {
+            assert len >= 2;
+
+            byte[] mask = new byte[4];
```

Comment:
Perhaps this mask byte array could be pre-allocated as a field instead of freshly instantiated for each frame?